src/icainfo.c

```diff
@@ -127,6 +127,74 @@ static int online_cca_card(void)
 	return 0;
 }
 
+int rsa_keylen_supported_by_openssl(unsigned int modulus_bitlength)
+{
+	unsigned char modexpo_public_e[512] = { 0 };
+	unsigned char modexpo_public_n[512] = { 0 };
+	unsigned char crt_private_p[256] = { 0 };
+	unsigned char crt_private_q[256] = { 0 };
+	unsigned char crt_private_dp[256] = { 0 };
+	unsigned char crt_private_dq[256] = { 0 };
+	unsigned char crt_private_inv_q[256] = { 0 };
+	ica_adapter_handle_t ah;
+	ica_rsa_key_mod_expo_t public_key;
+	ica_rsa_key_crt_t private_key;
+	int rc;
+
+	rc = ica_open_adapter(&ah);
+	if (rc != 0)
+		return 0;
+
+	public_key.modulus = modexpo_public_n;
+	public_key.exponent = modexpo_public_e;
+	public_key.key_length = (modulus_bitlength + 7) / 8;
+
+	private_key.p = crt_private_p;
+	private_key.q = crt_private_q;
+	private_key.dp = crt_private_dp;
+	private_key.dq = crt_private_dq;
+	private_key.qInverse = crt_private_inv_q;
+	private_key.key_length = (modulus_bitlength + 7) / 8;
+
+	rc = ica_rsa_key_generate_crt(ah, modulus_bitlength,
+							&public_key, &private_key);
+
+	ica_close_adapter(ah);
+
+	return rc == 0 ? 1 : 0;
+}
+
+/**
+ * Print out the minimum and maximum RSA key length. The maximum length is
+ * restricted to 4096 bits by crypto cards. The minimum accepted length in
+ * libica is 57 bits, but the available min length depends on the openssl
+ * version and fips mode.
+ */
+void print_rsa(void)
+{
+	int keylen_array[] = { 57, 512, 1024, 2048, 4096 };
```

Comment:
does anyone really care about an RSA key size of 57 bits ????
Even 512 bits is questionable.

Reply:
Yes indeed. I guess this is very old code in libica, but libica (and in fact openssl!) still support RSA with 57 bits. I would leave it up to applications whether to use such small key sizes, but found it worth to display the lower and upper boundaries. Note that the icainfo -r output just displays what openssl really allows. If openssl is configured to restrict small key sizes (e.g. via the fips provider/mode) then this output reflects this.